### PR TITLE
C#: Replace clears content with CSV summaries.

### PR DESCRIPTION
**Pull request by michaelnebel**
The shared library code already supports `WithoutContent`, which can be used to replace the override predicate `clearsContent`. Only parsing and printing functionality needed to be implemented for C# to leverage the shared withoutContent functionality.

That is, in this PR
- Implement the last small remaining C# language specific parts to leverage the withoutContent functionality.
- Implement the last small remaining C# language specific parts to leverage the withContent functionality.
- Replace all clearsContent predicates overrides with flow summaries. Note that this is practically infeasible due to the implementation of the existing clearsContent. An assumption was made in the original code that all classes that implements the `IEnumerable` interface and also has a `Clear` method, then the semantics of `Clear` would be that no flow is propagated via `Clear`. As a replacement, flow summaries for `Clear` has been introduced for many interfaces and classes in the .NET Runtime. For all practical reasons this is probably sufficiently good.
- Update flow summaries test cases.
- Delete the obsolete clearsContent framework code.

### CODE REVIEW

**Review by hvitved — COMMENTED** [review]
LGTM; just a small suggestion. We should also do a DCA run for C#.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImplSpecific.qll [hunk 1]
```diff
@@ -151,6 +151,8 @@ bindingset[c]
 SummaryComponent interpretComponentSpecific(AccessPathToken c) {
   c = "Element" and result = SummaryComponent::content(any(ElementContent ec))
   or
+  c = "WithoutElement" and result = SummaryComponent::withoutContent(any(ElementContent ec))
+  or
```
[the remark this answers is not in the document]
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/FlowSummaryImplSpecific.qll:155]: Sure thing!

**Review by hvitved — APPROVED** [review]

**Comment by michaelnebel** [comment]
Needed to rebase to execute DCA.

**Review by hvitved — DISMISSED** [review]

**Comment by michaelnebel** [comment]
@hvitved: Ran DCA twice yesterday. There a seems to some indication that some projects are slightly affected by performance (but not in alerts). Need to rebase again due to a rebase conflict. Will make a single threaded DCA execution with tuple counting this time around.

**Review by hvitved — APPROVED** [review]

**Comment by michaelnebel** [comment]
Finally DCA looks good
image

**Pull request merged**